generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -120,109 +249,44 @@ private Expr applyBinaryIdentityFunction(Expr expr, Expr opaqueExpr, BinOp opera
-          if (isZero) {
-            continue; // length(normalize(opaque)) only provides a means of representing 1, not 0
-          }
-          if (!BasicType.allGenTypes().contains(type)) {
-            continue; // normalize doesn't operate on non-gen types.
-          }
-          Expr normalizedExpr = new FunctionCallExpr("normalize", makeOpaqueZeroOrOne(isZero,
-              type, constContext, newDepth, fuzzer));
-          return new FunctionCallExpr("length", normalizedExpr);
-        case 4:
-          // represent 1 as cos(opaqueZero) function, e.g. cos(0.0)
-          if (isZero) {
-            continue; // cos(opaqueZero) only provides a means of representing 1, not 0
-          }
-          if (!BasicType.allGenTypes().contains(type)) {
-            continue; // cos doesn't operate on non-gen types.
-          }
-          return new FunctionCallExpr("cos", makeOpaqueZero(type, constContext, newDepth,
-              fuzzer));
-        case 5:
-          // represent 1 as the exponential function of opaqueZero, e.g. exp(0.0)
-          if (isZero) {
-            continue; // exp(opaqueZero) only provides a means of representing 1, not 0
-          }
-          if (!BasicType.allGenTypes().contains(type)) {
-            continue; // exp doesn't operate on non-gen types.
-          }
-          return new FunctionCallExpr("exp", makeOpaqueZero(type, constContext, newDepth,
-              fuzzer));
-        case 6:
-          // represent 0 as the natural logarithm of opaqueOne, e.g. log(1.0)
-          if (!isZero) {
-            continue; // log(opaqueOne) only provides a means of representing 0, not 1
-          }
-          if (!BasicType.allGenTypes().contains(type)) {
-            continue; // log doesn't operate on non-gen types.
-          }
-          return new FunctionCallExpr("log", makeOpaqueOne(type, constContext, newDepth,
-              fuzzer));
-        case 7:
-          // represent 0 as sin(opaqueZero) function, e.g. sin(0.0)
-          if (!isZero) {
-            continue; // sin(opaqueZero) only provides a mean of representing 0, not 1
-          }
-          if (!BasicType.allGenTypes().contains(type)) {
-            continue; // sin doesn't operate on non-gen types.
-          }
-          return new FunctionCallExpr("sin", makeOpaqueZero(type, constContext, newDepth,
-              fuzzer));
-        default:
-          throw new RuntimeException();
-      }
-    }
+
+    // If isZero holds, we are making an opaque zero, otherwise an opaque one
+    final List<OpaqueZeroOneFactory> opaqueFactories = isZero
+        ? OpaqueZeroOneFactory.waysToMakeZero(this)
+        : OpaqueZeroOneFactory.waysToMakeOne(this);
+    final List<Optional> availableOpaqueFactories = opaqueFactories
+        .stream()
+        .map(item -> item.tryMakeOpaque(type, constContext, newDepth, fuzzer, isZero))
+        .filter(Optional::isPresent)
+        .collect(Collectors.toList());
```

Comment:
The above is quite clever, but I do not like it ;-)  In particular, I don't like the strategy of trying to make an opaque expression to see whether it will be possible.

I would prefer something along the following lines, which loops forever until it finds something that works:

while (true) {
  final OpaqueZeroOneFactory factory = opaqueFactories.get(random index into opaqueFactories);
  final Optional<Expr> maybeResult = factory.tryMakeOpaque(...);
  if (maybeResult.isPresent()) {
    return maybeResult.get();
  }
}

By construction, there will always be at least one case that will work (injectionSwitch), so the loop will probabilistically terminate.

Reply:
Thank you! your code looks really clean and more understandable. Just out of my curiosity, why do we need `while (true)` in this case?

Reply:
The while (true) means that we keep on trying until we find a factory from which we get a non-empty optional (which we will eventually).

An alternative would be to loop while the list of factories is non-empty, and remove each unsuccessful factory from the list right after finding out the optional is not present. If the loop would terminate - indicating that nothing worked - a RuntimeException with appropriate error string should be thrown (rather than returning null) since we should always find a suitable factory I believe.

Reply:
The alternative approach is reasonable for me. The code has been fixed!
Thank you again for your informative comment.